cmd/trcd/main.go

```diff
@@ -29,31 +33,89 @@ func init() {
 func main() {
 	flag.Parse()
 
-	conn, err := net.Listen("unix", *sock)
+	var netConn net.Listener
+	var err error
+	if *useTCP {
+		log.Debug("Listening to TCP socket...")
+		addr, _ := net.ResolveTCPAddr("", ":4243")
+		netConn, err = net.ListenTCP("tcp", addr)
+
+	} else {
+		log.Debug("Listening to Unix socket...")
+		netConn, err = net.Listen("unix", *sock)
+	}
 	if err != nil {
 		panic(err)
 	}
-	defer conn.Close()
+	defer netConn.Close()
 
 	for {
-		c, err := conn.Accept()
+		sockConn, err := netConn.Accept()
+		log.Infof("Connected to %s", sockConn.RemoteAddr())
 		if err != nil {
 			panic(err)
 		}
-		defer c.Close()
+		defer sockConn.Close()
 
+		trcConn := Connect(sockConn, sockConn,
+			WithHandler(api.MessageTypeState, setStateHandler),
+			WithHandler(api.MessageTypePing, pingHandler),
+		)
+
+		// print errors resulting from mock trc
 		go func() {
-			dec := json.NewDecoder(c)
-			dec.DisallowUnknownFields()
-			var msg api.Message
-			if err := dec.Decode(&msg); err != nil {
-				log.Printf("Failed to decode input: %s", err)
+			for err := range trcConn.Errors() {
+				log.Error(err)
 			}
-			log.Printf("Got: %v", msg)
 		}()
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			log.Printf("Failed to copy input: %s", err)
+
+		// mock trc is running parallel to client anyway
+		if err := trcConn.SendHandshake(&api.Handshake{Version: trcapi.DefaultVersion}); err != nil {
+			panic(err)
+		}
+
+		// TODO initial state
+		if err := trcConn.SendState(&api.State{}); err != nil {
+			log.Error(err)
 			return
 		}
+
+		closeCh := make(chan struct{})
+		// TODO close when trc should shutdown
+		defer close(closeCh)
+
+		// send random messages unless specified
+		if !(*silent) {
+			go func() {
+				hBeat := time.Tick(time.Second)
+				for {
+					aWhile := time.Millisecond * time.Duration(rand.Intn(5000))
+					// wait 0-5 seconds, then send an empty state
+					select {
+					case <-time.After(aWhile):
+						if err := trcConn.SendState(&api.State{}); err != nil {
+							log.Error(err)
+						}
+					case <-hBeat:
+						trcConn.Ping()
+					case <-closeCh:
+						return
+					}
+				}
+			}()
+		}
 	}
 }
+
+func setStateHandler(msg *api.Message) (*api.Message, error) {
```

Comment:
can you make those anonymous or, (as a compromise) define them in the function itself, like so: